contracts/utils/Swap.sol

```diff
@@ -0,0 +1,454 @@
+     * @dev The nonce group is computed as nonce / 256, so each group of 256 sequential nonces uses the same key
+     * @dev The nonce states are encoded as 256 bits, for each nonce in the group 0 means available and 1 means used
+     */
+    mapping(address => mapping(uint256 => uint256)) internal _nonceGroups;
+
+    /************************************************
+     *  CONSTRUCTOR
+     ***********************************************/
+
+    constructor() {
+        uint256 currentChainId = getChainId();
+        DOMAIN_CHAIN_ID = currentChainId;
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                DOMAIN_TYPEHASH,
+                DOMAIN_NAME,
+                DOMAIN_VERSION,
+                currentChainId,
+                this
+            )
+        );
+    }
+
+    /************************************************
+     *  SETTER
+     ***********************************************/
+
+    /**
+     * @notice Sets the referral fee for a specific referrer
+     * @param referrer is the address of the referrer
+     * @param fee is the fee in percent in 2 decimals
+     */
+    function setFee(address referrer, uint256 fee) external onlyOwner {
+        require(referrer != address(0), "!referrer");
+        require(fee < MAX_PERCENTAGE, "Fee exceeds maximum");
+
+        referralFees[referrer] = fee;
+    }
+
+    /************************************************
+     *  OFFER CREATION AND SETTLEMENT
+     ***********************************************/
+
+    /**
+     * @notice Create a new offer available for swap
+     * @param offeredToken token offered by seller
+     * @param biddingToken token asked by seller
+     * @param minPrice minimum price of offeredToken in terms of biddingToken
+     * @param minBidSize minimum size allowed in terms of biddingToken
+     * @param totalSize amount of offeredToken offered by seller
+     */
+    function createOffer(
+        address offeredToken,
+        address biddingToken,
+        uint96 minPrice,
+        uint96 minBidSize,
+        uint128 totalSize
+    ) external override returns (uint256 swapId) {
+        require(offeredToken != address(0), "!offeredToken");
+        require(biddingToken != address(0), "!biddingToken");
+        require(minPrice > 0, "!minPrice");
+        require(minBidSize > 0, "!minBidSize");
+        require(totalSize > 0, "!totalSize");
+
+        offersCounter += 1;
+
+        swapId = offersCounter;
+
+        swapOffers[swapId] = Offer({
+            seller: msg.sender,
+            isOpen: true,
+            offeredToken: offeredToken,
+            biddingToken: biddingToken,
+            offeredTokenDecimals: IERC20Detailed(offeredToken).decimals(),
+            minBidSize: minBidSize,
+            minPrice: minPrice,
+            totalSize: totalSize,
+            availableSize: totalSize,
+            highestPrice: 0,
+            totalSales: 0
+        });
```

Comment:
A cheaper way to assign is to manually type it out like:

```solidity
swapOffers[swapId].seller = msg.sender;
swapOffers[offeredToken].seller = offeredToken;

// we don't have to assign totalSales because it defaults to 0
```

Reply:
The reason is that constructing the `Offer` struct in memory is cheaper